documentation/tutorials/kernel-debug.rst

```diff
@@ -0,0 +1,197 @@
+=========================
+  kw kernel debug tools
+=========================
+.. _kernel-debug:
+
+.. contents::
+   :depth: 1
+   :local:
+   :backlinks: none
+
+.. highlight:: console
+
+Introduction
+------------
+
+Linux Kernel provides multiple debug methods that can be handy to identify
+problems and learn how some specific feature works. With this idea in mind, kw
+tries to hide and standardize the complexity of dealing with the following
+debug features:
+
+* *Events*: Event tracepoints provide real-time information associated with some
+  specific parts of the code. Developers can enable this trace when the target
+  OS is running to collect data. If you want to know more about it, take a look
+  at: https://www.kernel.org/doc/html/latest/trace/events.html
+* Ftrace: Function callback traces can help identify some specific behavior
+  when a target function is invoked. In particular, the function graph filter
+  provides valuable information about the code sequence. If you want to learn
+  more about this feature, take a look at:
+  https://www.kernel.org/doc/html/v5.3/trace/ftrace-uses.html
+* Dmesg: In a few words, it shows the message buffer from the kernel.
+
+In particular, when we dive into events and ftrace we can abstract some common
+steps required to use these debug tools, which can be summarized as follows:
+
+#. Clean/Disable debugger::
+
+    echo 0 > /sys/kernel/debug/tracing/tracing_on
+
+#. Setup the target debug::
+
+    echo 'TARGET' > /sys/kernel/debug/tracing/[current_tracer|events...]
+
+#. Fine-tune the debug options::
+
+    echo 'FINE-TUNE' >> /sys/kernel/debug/tracing/[set_ftrace_filter|events/TARGET/filter]
+
+#. Enable the debug::
+
+    echo 1 > /sys/kernel/debug/tracing/tracing_on
+
+#. Collect the data::
+
+    cat /sys/kernel/debug/tracing/trace_pipe
+
+As you can see, we have multiple steps, and this situation can be worse when we
+consider that sometimes we want to collect these debug data in a machine
+connected to a local network or an external machine. Kw debug tries to hide
+this complexity under the "debug" option. For detailed information about this
+feature, check the man page, here we will provide a simple tutorial on how to
+use it.
+
+Common Options
+--------------
+
+The debug option provides a standard set of features that can be used with
+`--dmesg`, `--events`, and `--ftrace`. To avoid explaining the same ideas
+repeatedly, this section introduces the following shared behavior: `--follow`,
+`--history`, and `--cmd`.
+
+Follow log (`--follow | -f`)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to follow the log in real-time, you probably want to use the
+`--follow | -f` option, which will keep polling the data.
+
+Save (`--history | -k`)
+~~~~~~~~~~~~~~~~~~~~~~~
+
+This option will create a folder named `kw_debug` in the place that you run
+this command and a sub-folder that follow this pattern::
+
+  <ID>_<YEAR-MONTH-DAY>
+
+In a few words, every new execution is saved in that folder.
```

Comment:
Do the folders contain any files? If so, which ones? It isn't clear to me why I'd want to use this option.